# Key Learning Results

Key learning results:
- I understand the how to run Python code from the command line, in the interpreter, and in an Jupyter notebook.
- I can declare a variable.
- I feel comfortable writing various expressions.
- I understand "Pythonic" code organization.
- I recall the syntax for basic control flow statements.

# Where are we?: A brief introduction to running Python code

Before we actually start writing code, we need to talk about what we do with the code we're writing. Broadly, there are two ways of running Python code: line-by-line (a.k.a. interactively) or from a file. 

## Running code from file
Running a file in Python is as simple as openning up a terminal/command line window, navigating to the folder containing the file, then entering the following command:

```
python3 <file_name>
```

If you only have one version of Python installed or if the default version on Python on your system is Python 3, then it is also sufficient to use this command:

```
python <file_name>
```

Using the first command, however, ensures that our code is being run in Python 3.X and not Python 2.X. While Python 3.X is the most up-to-date, Python 2.X remains popular, and the two versions aren't perfectly cross compatible. This issue likely won't come up until you start exploring different, more advanced packages available in Python.

If you feel uncomfortable navigating your computer from the terminal/command line, I suggest [working through this short tutorial.](https://blog.teamtreehouse.com/introduction-to-the-mac-os-x-command-line) You don't need to become an expert, but a basic understanding of how to interface with your computer through the command line.

Programs don't by default have a command line interface, and though building one out is fairly simple, it won't be covered in any of the Python Jams. Still, it's good to be aware of how to run programs from the command line.

## Running code interactively

Throughout these sessions, we'll be running code interactively either through IPython or through a Jupyter notebook. Both allow us to work through code line by line, evaluating our progress as we work. 

### IPython

To access IPython, open up the terminal/command line and run this command:

```
ipython3
```

You should see the following text appear on screen:

```
Python 3.7.3 (default, Mar 27 2019, 16:54:48) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.5.0 -- An enhanced Interactive Python. Type '?' for help.
```

This text is followed by a prompt where you can type in Python code and see its results.

### Your first program

Now that you're ready to write Python code, lets print out a simple Hello World statement. In Python, this requires a single line of code:

```
print("Hello, world!")
```

Type this into the IPython window we just opened, and you should see the text ```Hello, world!``` appear below where you just typed!

### Jupyter notebooks

Another place where we can run code iteractively is a Jupyter notebook. Indeed, this document you're reading right now is a Jupyter notebook. Running code in a Jupyter notebook allows you to combine text (in Markdown), plots, and other media alongside your code. This ability can be helpful if you're using Python to produce an analysis you intend to send to others or if you would like to explain the code you're writing as you're writing it. Furthermore, Jupyter notebooks also allow us to export 

Below this text, there is an empty "code" cell where you can write and run Python code within this document. Try typing our Hello, World! statement in that cell and then run it by either pressing Shift + Return or pressing the run button in the navigation bar.

In [48]:
print('Hello, World!')

Hello, World!


The majority of Python Jams will be in Jupyter notebooks, however, you should also try running some code in IPython so that you feel comfortable with it. When working by yourself on a project, you'll likely find it easier to work in IPython than in Jupyter notebooks.

# Basics of Python Programming

The topics covered in this section may seem elementary, but are essential building blocks for writing or debugging more advanced programs. While the section will be focused on how these topics are implemented in Python, these elements will be available in any programming lanague, though possibly with a slightly different impelementation.

## Variables

Variables are symbolic links that allow you to keep track of information in your machine's memory while a program runs. When defining a variable, you give it a name that you will use later on to retrieve or update the information stored in that variable.

In Python, we define variables like this:

```
<variable name> = <value>
```

Let's try rewriting our Hello, World! statement above using a variable. First, define a variable named *message* with the value "Hello, World!" in the empty cell below then press Shift + Return. Notice that when you want to store a block of text (called a string) as a variable, you have to wrap the text block in quotation marks. In Python, you can use either single (') or double (") quotation marks to wrap strings.

In [49]:
message = "Hello, World!"

Now, we can use the print function from above to print out the value of our *message* variable with the command

```
print(message)
```

Try it in the empty cell below:

In [50]:
print(message)

Hello, World!


You can also check the value of a variable by simply typing its name into the interpreter. This is helpful when working on or debugging a program but will not produce any output if you're running a program from the command line. Check the value of our message variable below:

In [51]:
message

'Hello, World!'

## Data Types

If you've ever written code in another language, you may have felt like something was missing in our variable assignment statement above. Namely, in Python there is no need to declare the type of variable you're declaring before you declare it.  

What does this mean? In Java, for example, you would declare the variable above as"

```
String message = "Hello, World!"
```

That first keyword, "String", is used to let Java know that the data you're storing is a string of text. Python, however, doesn't require explicity stating what type of data you're storing in a variable when you declare it. This doesn't mean, however, that variables in Python don't have a specific type associated with them. Indeed, we can check the type of a variable by using the type function.

In [52]:
type(message)

str

"str" means that this variable is a string. In Python, there are four basic types of variables and number of subtypes. The most important ones to know are:  

- Numerics: Can encompass both integer and real numbers
    - Integers (exclusively for whole numbers)
    - Floats/Floating point decimals (for whole numbers or decimals)

In [53]:
x = 5
type(x)

int

In [54]:
y = 11.012017
type(y)

float

In [55]:
type(y)

float

- Strings: Represent text as a 'string' of characters (i.e. the comptuer sees 'Hello, World!' as an 'H' character followed by an 'e' character followed by an 'l' character, etc.). There are a number of subtypes of strings with specific applications that will be covered later.

In [56]:
text = 'Welcome to Python!'
type(text)

str

- Booleans: represent logical values, i.e. True or False

In [57]:
t_or_f = True
type(t_or_f)

bool

- "None": Indicates the absence of a value. In some cases, it's useful to be able to explicitly note that a value does not exist or was not provided by a user.

In [58]:
a = None
type(a)

NoneType

Now, try declaring a few variables for youself and checking their types using the ```type()``` function.

## Expressions

Now that you know how to declare variables and have seen some of the different types of variables, you're ready to start combining them to produce new values.  

### Arithmetic

One extremely basic use for Python would be as a simple calculator. All the basic arithmetic expressions, addition (```+```), subtraction (```-```), multipliation (```*```), exponentiation (```**```)division (```/```) work basically as expected.

In [59]:
11 + 1

12

In [60]:
11 - 1

10

In [61]:
11 * 2

22

In [62]:
11 ** 2

121

In [63]:
12 / 2

6.0

In the last answer, notice that we divided one integer by another yet the output was a float (you can tell by the decimal point). Standard division in Python will always produce a float, regardless of the two values you're dividing. If for some reasons you'd rather have an integer result, you should use the integer division operator, ```//```.

In [64]:
12 // 2

6

In [65]:
11 // 2

5

In [66]:
-11 // 2

-6

In effect, integer division will divide the two values then round down to the nearest integer.  

Lastly, there is also a modulus operator, ```%```, that yields the remainder of an integer division operation.

In [67]:
3.3 % 3

0.2999999999999998

<div class="alert alert-block alert-warning">
<b>Whoa, isn't that answer wrong?</b><br />
In the above cell block, you may have noticed that Python returns a value of 0.2999999999999998 instead of 0.3. This issue occurs because of how computers must use an approximation to store decimals. In some scientific operations, these so-called "floating point errors" can cause an issue, and it's necesary to use the Python decimal type. If you're curious about this, I'd encourage you to <a href="https://docs.python.org/2/library/decimal.html">read the Python documentation on decimal types.</a>
</div>  



We can also chain operators using the standard mathematical order of operations.

In [68]:
10 * 2 + 2

22

Finally, we can also use variables in our expression either to store the result of our expressions or perform operations on our different varaibles.

In [69]:
a = 10
b = 2
a * b + b

22

In [70]:
c = a * b + b

In [71]:
print(c)

22


### Strings

Python's built-in operators also provide an easy was to join two strings together using the addition operator.

In [72]:
'abc' + 'def'

'abcdef'

In [73]:
message1 = 'Hello '
message2 = 'world!'
message3 = message1 + message2
print(message3)

Hello world!


Notice, however, that attempting to add a string and a number will yield an error.

In [74]:
'one' + 1

TypeError: can only concatenate str (not "int") to str

### Relational operators

This next set of operators are used to compare two values, and the output of an expression involving them is always a boolean. The include:

- ```>```: greater than
- ```>=```: greater that or equal to
- ``` <```: less than
- ```<=```: less than or equal to
- ```==```: equals
- ```!=```: not equals

All six of these can be used on numeric type data, and the final two (equals and not equals) can be used on string type data. Let's see them in action:

In [75]:
5 == 5

True

In [76]:
10 < 2

False

In [77]:
6 != 7

True

In [78]:
"Hello" == "Hello"

True

In [79]:
"Hello" == "Bonjour"

False

In [80]:
"hello" == "Hello"

False

Notice that string equality is case sensitive.

In the cell below, I've defined two variables, ```a``` and ```b```. Write three expressions in the empy cells to check wheter a is greater than zero, whether b is greater than or equal to zero, and whether a times b is less than zero.

In [81]:
a = 10
b = -1

In [82]:
#is a greater than zero? enter your code below this line

In [83]:
#is b greater than or equal to zero? enter your code below this line

In [84]:
#is a + b equal to zero? enter your code below this line

### Logical operators

We can combine multiple expressions with relational operators using the logical operators ```and``` and ```or```. Consider this expression:

In [85]:
a = 20
b = 30

a > 25 and b > 25

False

This expression would only evaluate to True if both a and b were greater than 25.

In [86]:
a > 25 or b > 25

True

This expression will evaluate to True if either a or b was greater than 25. 

Finally, there is also a ```not``` operator that will negate a boolean value. For example:

In [87]:
a = 5
a > 0

True

In [88]:
not a > 0

False

The value of ```a``` is greater than zero, therefore the first expression evaluate to True. The not operator, however, negates this True value in the second expression, so the second expression produes False. Consider another example. Do you understand why the expression evaluates to be True?

In [89]:
a = 5
not a < 0

True

Similar to arithmetic, you can (and should!) use parentheses in your expressions to clarify order of operations. The next expression checks that a is either less than or equal to zero or greater than or equal to 5.

In [90]:
not (a > 5 or 0 < a)

False

The following expression will check whether a is greater than 0 or less than or equal to 5.

In [91]:
 not a > 5 or a < 0

True

# Control Flow

So far, the programs we've written have been fairly simple and boring. Print a message, calculate a sum, etc. Control flow statements allow us to start writing more interesting programs. The three most commong types of control flow statements that you'll wee are:  

- Conditional statements, or "if"/"if-then-else statements," run a piece of code only if a certain condition is true.
- Sequence-based loops, or "for loops," run a piece of code repeatedly based on a set number of repetitions.
- Condition-based loops, or "while loops," run a piece of code repeatedly so long as a boolean condition is true.

## If Statements

For the first example, lets check with a number is even or odd.

In [92]:
n = 7
if n % 2 == 1:
    print(n, 'is odd')
else:
    print(n, 'is even')

7 is odd


There are a few points to note in the previous example. First, lets look at the syntax of an if statement in python. They have the general form:  

```
if <boolean expression>:
    <code to run if true>
```

We can also introduce variants of the above syntax that allow us to specify what to do if the boolean expression is false (as we did in the example).

```
if <boolean expression>:
    <code to run if true>
else:
    <code to run if false>
```

Finally, we can also use a variant that allows us to check multiple conditions.

```
if <boolean expression #1>:
    <code to run if expression #1 is true>
elif <boolean expression #2>:
    <code to run if expression #2 is true>
elif <boolean expression #3>:
    <code to run if expression #3 is true>
...
...
...
...
else:
    <code to run if all expressions are false>
```

It's important to note, however, for any if statement a maximum of one code block will run. More specifically, the first code block following a true boolean expression will run and then the interpreter will exit the statement. For example:

In [93]:
n = 6
if n % 2 == 0:
    print(n, 'is divisible by 2')
elif n % 3 == 0:
    print(n, 'is divisible by 3')

6 is divisible by 2


Despite the fact that n is divisible by both two and three, the program above only prints out that 6 is divisible by 2 because that is the first true boolean expression it finds. The correct way to write the above program would be to use multiple if statements, rather than an if statment with multiple branches.

In [94]:
n = 6
if n % 2 == 0:
    print(n, 'is divisible by 2')
if n % 3 == 0:
    print(n, 'is divisible by 3')

6 is divisible by 2
6 is divisible by 3


The second item worth noticing is how Python distinguishes between code blocks. While many other programming languages use curly braces ```{``` and ```}``` to distinguish between code blocks, Python uses indentation. Though you can technically pick any consistent indentation pattern to indicate a new block, the cannonical pattern is using four spaces to correspond to one level of indentation.  

Python is incredibly picky about this which an at times result in some frustrting errors. For example:

In [95]:
condition = True
if condition:
    print('The condition is true!')
	print('End of if statement')

TabError: inconsistent use of tabs and spaces in indentation (<ipython-input-95-c6b2aeff1d66>, line 4)

On it's face, it looks like the above code should run without issue -- the indentation pattern appears to be the same. However, in the first line of code block four spaces are used to delimit the codeblock, but in the second line of the code block, a single tab character is used to delimit the codeblock. This issue is particularly common if you used multiple machines or multiple text editors on the same file. 

The third and final item worth noticing is that these print statements look different than what we were using before. In general, if we pass multiple values to a print statement and separte them by a comma, then all the values will be printed out in the order they were passed separated by a space. This can be super helpful and save you from having to manage string concatenations when printing out a value.

Here is an example for you to try:
- Write a conditional statement to check whether a number, ```n``` is positive, negative, or neither (i.e. equal to zero)

After you write your program, modify the value of n to check that your statement is correct.

In [96]:
n = 17
### your code here

## For loops

For loops are a way to dynamically repeat code and avoid typing the same statement over an over again. We'll revisit for loops more talking about lists and other data structures, but for now it's enough for you to know that the basic syntax of a for loop in Python is:

```
for <variable name> in <sequence of values>:
    <code to execute>
```

Let's look at this in practice. Let's pretend we're the owners of a thrift shop and our pricing scheme is to take the full price of an item and charge 12.5% of that full price. Suppose we have 10 items, with full prices $10, $25, $16, $17, $85, $35, $67, $75, $55, and $42. One way to use Python to solve this problem would be to write a separate statement for each price. 

In [97]:
print(10*.125)

1.25


In [98]:
print(25*.125)

3.125


In [99]:
print(16*.125)

2.0


You can quickly see how this could become tedious. Another way to do this would be to use a for loop.

In [100]:
for price in [10, 25, 16, 17, 85, 35, 67, 75, 55, 42]:
    print(price * .125)

1.25
3.125
2.0
2.125
10.625
4.375
8.375
9.375
6.875
5.25


Much faster (and much less typing)! Notice that we wrapped our sequence of prices in brackets; this item is called list and will be a major focus of the next Python Jam.  

Let's talk about another use of for loops: doing something for every whole number in a range. For example, let's say I want to list out whether every number between 1 and 100 is divisible by 3. One way to do this would be to type out every number between 1 and 100. Thankfully, however, Python provides a function, called ```range()``` that will do this for us. To get every number between 1 and 100, we would use the function like this:

In [101]:
for i in range(1, 101):
    if i % 3 == 0:
        print(i, 'is divisible by 3')
    else:
        print(i, 'is NOT divisible by 3')

1 is NOT divisible by 3
2 is NOT divisible by 3
3 is divisible by 3
4 is NOT divisible by 3
5 is NOT divisible by 3
6 is divisible by 3
7 is NOT divisible by 3
8 is NOT divisible by 3
9 is divisible by 3
10 is NOT divisible by 3
11 is NOT divisible by 3
12 is divisible by 3
13 is NOT divisible by 3
14 is NOT divisible by 3
15 is divisible by 3
16 is NOT divisible by 3
17 is NOT divisible by 3
18 is divisible by 3
19 is NOT divisible by 3
20 is NOT divisible by 3
21 is divisible by 3
22 is NOT divisible by 3
23 is NOT divisible by 3
24 is divisible by 3
25 is NOT divisible by 3
26 is NOT divisible by 3
27 is divisible by 3
28 is NOT divisible by 3
29 is NOT divisible by 3
30 is divisible by 3
31 is NOT divisible by 3
32 is NOT divisible by 3
33 is divisible by 3
34 is NOT divisible by 3
35 is NOT divisible by 3
36 is divisible by 3
37 is NOT divisible by 3
38 is NOT divisible by 3
39 is divisible by 3
40 is NOT divisible by 3
41 is NOT divisible by 3
42 is divisible by 3
43 is NOT divis

Notice that the lower bound of the range function is inclusive while the upper bound is not.

Now, try writing a program using a for loop to calculate 9! (i.e. 9 x 8 x 7...x 1):

In [102]:
result = 1
for num in range(1, 10):
    result = result * num
print(result)

362880


## While loops

While loops are a more simple type of loop that repeats an action so long as a condition remains true. The general syntax for a while loop is:

```
while <boolean expression>:
    <code block>
```

Let's try rewriting our program to calculate 9! using a for loop.

In [103]:
N = 9
i = 1
result = 1

while i <= N:
    result = i * result
    i = i + 1

print(result)

362880


Right now, while loops may simply seem like less clean for loops, however, as you start doing more advanced computations while loops will begin to become more useful.